internal/app/app_test.go

```diff
@@ -0,0 +1,88 @@
+package app_test
+
+import (
+	"context"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/require"
+	"github.com/ubuntu/ubuntu-proxy-manager/internal/app"
+	"github.com/ubuntu/ubuntu-proxy-manager/internal/testutils"
+)
+
+func TestApp(t *testing.T) {
+	tests := map[string]struct {
+		applyArgs    []string
+		noSystemBus  bool
+		noMethodCall bool
+
+		wantNewErr      bool
+		wantDbusCallErr bool
+		wantRetCode     int
+	}{
+		"Empty apply arguments":   {},
+		"Correct apply arguments": {applyArgs: []string{"http://proxy:3128", "", "", "", "", ""}},
+		"No method called on app": {noMethodCall: true},
+
+		"Error when system bus is not available": {noSystemBus: true, wantNewErr: true},
+	}
+
+	for name, tc := range tests {
+		tc := tc
+		name := name
+		t.Run(name, func(t *testing.T) {
+			if !tc.noSystemBus {
+				defer testutils.StartLocalSystemBus()()
+			}
+
+			if len(tc.applyArgs) == 0 {
+				tc.applyArgs = []string{"", "", "", "", "", ""}
+			}
+
+			args := make([]interface{}, len(tc.applyArgs))
+			for i := range tc.applyArgs {
+				args[i] = tc.applyArgs[i]
+			}
+
+			ctx := context.WithValue(context.Background(), app.DryRun{}, true)
+			a, err := app.New(app.WithInitCtx(ctx), app.WithTimeout(50*time.Millisecond))
+			if tc.wantNewErr {
```

Comment:
I would separate the `TestNew()` and the `TestApply()` part as here, it’s easy to make one failing and not the other one. That way, the matrix of test cases are only scoped the specific matter.